## Задача 1

Контрольная работа состоит из пяти вопросов. На каждый вопрос приведено четыре варианта ответа, один из которых правильный. Составьте закон распределения числа правильных ответов при простом угадывании. Найдите M(X), D(X).

#### Решение:

Принимаемые значения данной случайной величины будут: $x_1 = 0$, $x_2 = 1$, $x_3 = 2$, $x_4 = 3$, $x_5 = 4$, $x_6 = 5$

Воспользуемся формулой Бернулли для нахождения вероятности каждого события. 

In [1]:
import numpy as np

In [2]:
def bernoulli_distr(n: int, k: int, p: float) -> float:
    
    return (np.math.factorial(n) * np.power(p, k) * np.power(1 - p, n - k)) / (np.math.factorial(k) * np.math.factorial(n - k))

Для каждого билета вероятность попадания случайным образом в правильный ответ равна $P(X) = \dfrac{1}{4}$.

In [3]:
distr_a = []
for i in range(6):
    distr_a.append(bernoulli_distr(5, i, 0.25))
print(distr_a)

[0.2373046875, 0.3955078125, 0.263671875, 0.087890625, 0.0146484375, 0.0009765625]


Закон распределения:

<table border="3">
    <tr>
        <th>$x$</th>
        <td>0</td>
        <td>1</td>
        <td>2</td>
        <td>3</td>
        <td>4</td>
        <td>5</td>
</tr>
    <tr>
        <th>$P(X=x)$</th>
        <td>0.237</td>
        <td>0.396</td>
        <td>0.264</td>
        <td>0.088</td>
        <td>0.015</td>
        <td>0.001</td>
    </tr>
</table>

Чтобы найти математическое ожидание, поэлементно перемножим значения двух массивов, затем сложим результаты.

In [4]:
x_values = np.array([0, 1, 2, 3, 4, 5])
x_probabilities = np.array(distr_a)

In [5]:
m = x_values.dot(x_probabilities)

print(m)

1.25


Определим дисперсию случайной величины $X$. 

Найдем случайную величину $Y = X - M(X)$. Она будет принимать значения:

In [6]:
y_values = x_values - m

print(y_values)

[-1.25 -0.25  0.75  1.75  2.75  3.75]


Далее, возведем эту величину в квадрат, тем самым, получим новую случайную величину  𝑍  со все еще таким же распределением вероятностей:

In [7]:
z_values = y_values ** 2

print(z_values)

[ 1.5625  0.0625  0.5625  3.0625  7.5625 14.0625]


Итак, дисперсия случайной величины  𝑋  равна:

In [8]:
d = z_values.dot(x_probabilities)

print(d)

0.9375


## Задача 2

Вероятность попадания в цель при одном выстреле равна 0.01. Какова вероятность того, что число попаданий при 200 выстрелах будет не менее 5 и не более 10?

#### Решение:

Воспользуемся распределением Пуассона, поскольку имеется большое количество событий. 

In [121]:
from decimal import *
def poisson_distr(a, k):
    
    return (Decimal(a ** k) * Decimal(np.exp(-a)))/ Decimal(np.math.factorial(k))

Найдем вероятность попадания в цель для каждого случая 5-10 выстрелов, и, поскольку события несовместные, сложим полученные вероятности:

In [122]:
n = 200

In [123]:
p = 0.01

In [124]:
P5_10 = 0

In [125]:
for i in range(5, 11):
    P5_10 = P5_10 + poisson_distr(n * p, i)

In [126]:
print(P5_10)

0.05264470911934267658415695911


## Задача 3

В условиях предыдущей задачи сколько выстрелов нужно сделать чтобы быть уверенным с вероятностью 0.9, что хотя бы 10 раз будет совершено попадание?

Воспользуемся формулой из предыдущего примера и подставим её в цикл, пока не достигнем нужной вероятности 0.9. При это будем игнорировать ничтожно малые величины $<10^{-5}$:

In [162]:
def distr_sum(n):
    psum = 0
    for i in range(10, n + 1):
        psum += poisson_distr(n * p, i)
        if poisson_distr(n * p, i) < 10**(-5):
            break
    return psum

In [163]:
n = 10
while distr_sum(n) < 0.9:
    n+=1 
print(n)

1421
